### PR TITLE
feat(core): add support for custom prompts #9038

### LINKED ISSUE #9038: Add support for other enquirer prompts.
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
[Enquirer](https://github.com/enquirer/enquirer) is used by `@nrwl/tao` to [handle the prompts for generator options via the schema](https://github.com/nrwl/nx/blob/5a74c02a49076a145cdba8d475b03193f317c889/packages/tao/src/shared/params.ts#L682-L748). Currently only the [`input`](https://github.com/enquirer/enquirer#input-prompt), [`select`](https://github.com/enquirer/enquirer#select-prompt), [`multiselect`](https://github.com/enquirer/enquirer#multiselect-prompt), [`confirm`](https://github.com/enquirer/enquirer#confirm-prompt), & [`numeral`](https://github.com/enquirer/enquirer#numeral-prompt) prompts are supported.

However, there are several other prompt types that would be very convenient to have for some generators, such as
- [`list`](https://github.com/enquirer/enquirer#list-prompt) prompts for when you want an array of items, but don't want to constrain the options, like `multiselect` does
- [`autocomplete`](https://github.com/enquirer/enquirer#autocomplete-prompt) prompts for when you want to suggest options, but not constrain input to them, like `select` does
- [`invisible`](https://github.com/enquirer/enquirer#invisible-prompt) or [`password`](https://github.com/enquirer/enquirer#password-prompt) might be good for secrets, if you have a generator that needs to access cloud resources/ private repos.
- less obvious how it would be implmented, but [`snippet`](https://github.com/enquirer/enquirer#snippet-prompt) prompts would provide a very nice & clean way to make templating for generators that use EJS templates interactive

## Motivation
I initially thought of this because I wanted to create a generator that would be able to add targets for running `@nx-tools/nx-docker` to my apps. I was starting to lay it out and I realized that for a `tags` input, there's no effective way to prompt for a free-form list, either I make the schema option demand a single delimited string, and then manually split it in the generator's code, or I don't provide a prompt. The fact that enquirer provides a `list` prompt that would be about a 5-line implementation was too tantalizing to ignore.

After looking at enquirer, I realized that several other prompt types might lend themselves very nicely to nx's patterns (described above)

## Suggested Implementation
The `list` prompt type should be very straightforward to implement, as would the `autocomplete`, `invisible`, or `password` prompts; just pass through the prompt type (if it's one of those) and add some checks to validate that the type of the property matches what the prompt type returns.

## Alternate Implementations
The `snippet` prompt for doing interactive templating would require significantly more effort, and may not be feasible, but I imagining that it could probably be implemented look at the options allowed by the schema and then create a substitutions object to be used by EJS to render the template file into a template string that you can be passed to the snippet prompt. The problem here lies in the fact that the snippet prompt only returns the single, rendered template string and doesn't have support for multiple templates or surfacing only what the filled in values were.

To have this jive well with the nx cli approach, you'd need to write a custom prompt based on the approach of the snippet prompt. It would need to be made aware of all of template files, maybe showing them all at once at once (ugly), iterate over them, or add some additional interactions to toggle between templates that still need to be filled in. You'd also want to have it return just the object of options as filled in by the user, so as not to disrupt the current flow of the generators (i.e. collect/resolve the prompted options & only then run generator logic – e.g. calling `@nrwl/devkit`'s `generateFiles()`).

### PR DESCRIPTION
**Pull request by nzacca**
## Current Behavior
* A subset of Enquirer prompts may be used within Nx workspace generators.

## Expected Behavior
* [Built-in prompts](https://github.com/enquirer/enquirer#-built-in-prompts) supported by Enquirer may now be used within Nx workspace generators.

## Related Issue(s)

Fixes #9038

### CODE REVIEW

**Comment by nzacca** [comment]
@FrozenPandaz 

Note this PR reverted [a recent change](https://github.com/nrwl/nx/pull/11737) you made to the `getPromptsForSchema` function by making the function `async` once again. This was done in order to allow for `async` custom prompts. I hope this is OK but please let me know if there were other driving factors behind removing `async` in that commit.

**Comment by luxaritas** [comment]
It would be awesome to support prompts that depend on previous values! Can this be adjusted so that custom prompts are not necessarily in the tools directory, so that custom prompts could work with Nx plugins?

**Comment by nzacca** [comment]
@luxaritas Thank you for your feedback. 

This PR opens up the entire enquirer lib for Nx so obtaining previous values is already supported via the `this.state.answers` property in a prompt.

As for referencing custom prompts exposed by a package (vs a local prompt), I will look into adding support for that.

**Comment by luxaritas** [comment]
FYI, it would be important to support both resolving from node_modules (for installed plugins) as well as local projects (for local plugins)

**Review by FrozenPandaz — COMMENTED** [review]
Custom prompts will take some more thinking about implementation. It's likely best for @AgentEnder to implement it when he has time.

The original issue asks for `list`, `invisible`, and `password` which I think would be better to start with. Let's do 1 at a time. Let's discuss exactly how nx plugin developers would write their schemas to get a type of prompt. You can then open PRs after it's agreed upon.

File: packages/nx/src/utils/params.ts [hunk 1]
```diff
@@ -759,15 +772,43 @@ export function getPromptsForSchema(
...
       } else if (
-        v['x-prompt'].type == 'confirmation' ||
-        v['x-prompt'].type == 'confirm'
+        v['x-prompt'].type === 'confirmation' ||
+        v['x-prompt'].type === 'confirm'
       ) {
         question.message = v['x-prompt'].message;
         question.type = 'confirm';
+      } else if (isCustomPrompt(v['x-prompt'])) {
+        const toolsOutDir = getToolsOutDir();
```
[the remark this answers is not in the document]
Comment (nzacca) [reply packages/nx/src/utils/params.ts:785]: @FrozenPandaz 

Thank you for the feedback. While this does work for locally defined prompts, I did not consider the use case of published packages/local plugins. I've left more feedback on the issue.

**Pull request closed without merging**